File: frontend/src/lib/gemini-client.ts
/**
 * Gemini AI Client for DataLens
 * Provides AI-powered insights and recommendations for EDA
 */

import { GoogleGenerativeAI } from '@google/generative-ai'
import type { AnalysisResult } from './data-processor'

// Initialize Gemini AI
const genAI = new GoogleGenerativeAI(import.meta.env.VITE_GEMINI_API_KEY || '')
const model = genAI.getGenerativeModel({ model: 'gemini-pro' })

export interface TransformationSuggestion {
  column: string
  transformation: string
  reason: string
  code?: string
}

export interface DataQualityFlag {
  severity: 'low' | 'medium' | 'high' | 'critical'
  category: 'missing_values' | 'outliers' | 'inconsistency' | 'correlation' | 'distribution'
  message: string
  recommendation: string
}

export interface MLRecommendation {
  model: string
  problemType: 'classification' | 'regression' | 'clustering' | 'dimensionality_reduction'
  rationale: string
  preprocessing: string[]
  expectedPerformance: string
  considerations: string[]
}

export interface PatternInsight {
  pattern: string
  description: string
  confidence: number
  implications: string[]
}

/**
 * Generate comprehensive narrative insights using Gemini AI
 */
export async function generateNarrativeInsights(
  analysisResult: AnalysisResult,
  datasetName: string
): Promise<string> {
  try {
    const prompt = `You are an expert data scientist analyzing a dataset called "${datasetName}".

Based on the following statistical analysis, provide a 2-paragraph executive summary:

Dataset Overview:
- Total Rows: ${analysisResult.summary.totalRows.toLocaleString()}
- Total Columns: ${analysisResult.summary.totalColumns}
- Numerical Columns: ${Object.keys(analysisResult.statistics.numerical).length}
- Categorical Columns: ${Object.keys(analysisResult.statistics.categorical).length}

Key Statistics:
${Object.entries(analysisResult.statistics.numerical).slice(0, 5).map(([col, stats]) =>
  `- ${col}: mean=${stats.mean.toFixed(2)}, std=${stats.std.toFixed(2)}, range=[${stats.min}, ${stats.max}]`
).join('\n')}

Correlations:
${analysisResult.correlations.slice(0, 5).map(c =>
  `- ${c.col1} vs ${c.col2}: ${c.correlation.toFixed(3)}`
).join('\n')}

Data Quality:
- Missing Values: ${Object.values(analysisResult.summary.missingValues).reduce((a, b) => a + b, 0)}
- Missing Percentage: ${((Object.values(analysisResult.summary.missingValues).reduce((a, b) => a + b, 0) / (analysisResult.summary.totalRows * analysisResult.summary.totalColumns)) * 100).toFixed(1)}%

Provide:
1. A concise overview of the dataset's characteristics and key patterns
2. Critical data quality observations and their potential impact
3. Brief mention of the most interesting relationships found

Keep it professional, actionable, and under 200 words total.`

    const result = await model.generateContent(prompt)
    const response = await result.response
    return response.text()
  } catch (error) {
    console.warn('Gemini API error, using fallback:', error)
    return generateFallbackNarrative(analysisResult)
  }
}

/**
 * Generate transformation suggestions using Gemini AI
 */
export async function suggestTransformations(
  columnName: string,
  stats: any,
  analysisResult: AnalysisResult
): Promise<TransformationSuggestion[]> {
  try {
    const prompt = `As a data scientist, recommend data transformations for column "${columnName}" with these statistics:

- Mean: ${stats.mean?.toFixed(2) || 'N/A'}
- Median: ${stats.median?.toFixed(2) || 'N/A'}
- Std Dev: ${stats.std?.toFixed(2) || 'N/A'}
- Min: ${stats.min || 'N/A'}
- Max: ${stats.max || 'N/A'}
- Missing Values: ${analysisResult.summary.missingValues[columnName] || 0}

Recommend specific transformations to improve normality and ML model performance.
Provide 2-3 suggestions in JSON format:
{
  "transformations": [
    {
      "transformation": "log transform",
      "reason": "reduces right skew",
      "code": "np.log(column)"
    }
  ]
}

Focus on practical, commonly-used transformations.`

    const result = await model.generateContent(prompt)
    const response = await result.response

    // Try to parse JSON response
    try {
      const parsed = JSON.parse(response.text())
      return parsed.transformations || []
    } catch {
      // If JSON parsing fails, extract suggestions textually
      return parseTextualSuggestions(response.text(), columnName)
    }
  } catch (error) {
    console.warn('Gemini transformation suggestion error:', error)
    return generateFallbackTransformations(columnName, stats)
  }
}

/**
 * Detect data quality issues using Gemini AI
 */
export async function detectDataQualityIssues(
  analysisResult: AnalysisResult
): Promise<DataQualityFlag[]> {
  try {
    const issues: DataQualityFlag[] = []

    // Check for high missing values
    Object.entries(analysisResult.summary.missingValues).forEach(([col, missing]) => {
      const missingPercent = (missing / analysisResult.summary.totalRows) * 100
      if (missingPercent > 30) {
        issues.push({
          severity: 'critical',
          category: 'missing_values',
          message: `"${col}" has ${missingPercent.toFixed(1)}% missing values`,
          recommendation: `Consider imputation (median/mode) or removing this column`
        })
      } else if (missingPercent > 15) {
        issues.push({
          severity: 'high',
          category: 'missing_values',
          message: `"${col}" has ${missingPercent.toFixed(1)}% missing values`,
          recommendation: `Apply imputation strategies or remove affected rows`
        })
      } else if (missingPercent > 5) {
        issues.push({
          severity: 'medium',
          category: 'missing_values',
          message: `"${col}" has ${missingPercent.toFixed(1)}% missing values`,
          recommendation: `Monitor impact on analysis, consider imputation`
        })
      }
    })

    // Check for high correlations
    analysisResult.correlations.forEach(corr => {
      if (Math.abs(corr.correlation) > 0.95) {
        issues.push({
          severity: 'high',
          category: 'correlation',
          message: `Very high correlation (${corr.correlation.toFixed(3)}) between "${corr.col1}" and "${corr.col2}"`,
          recommendation: `Consider removing one variable to avoid multicollinearity`
        })
      } else if (Math.abs(corr.correlation) > 0.85) {
        issues.push({
          severity: 'medium',
          category: 'correlation',
          message: `High correlation (${corr.correlation.toFixed(3)}) between "${corr.col1}" and "${corr.col2}"`,
          recommendation: `Monitor for multicollinearity in models`
        })
      }
    })

    // Check for extreme outliers in numerical data
    Object.entries(analysisResult.statistics.numerical).forEach(([col, stats]) => {
      const iqr = stats.quartiles[2] - stats.quartiles[0]
      const outlierThreshold = stats.quartiles[2] + (3 * iqr)

      if (stats.max > outlierThreshold) {
        issues.push({
          severity: 'medium',
          category: 'outliers',
          message: `"${col}" contains potential outliers (max: ${stats.max})`,
          recommendation: `Consider capping or transforming extreme values`
        })
      }
    })

    return issues
  } catch (error) {
    console.warn('Data quality detection error:', error)
    return []
  }
}

/**
 * Recommend ML models using Gemini AI
 */
export async function recommendMLModels(
  analysisResult: AnalysisResult,
  targetVariable?: string
): Promise<MLRecommendation[]> {
  try {
    const numericalCols = Object.keys(analysisResult.statistics.numerical)
    const categoricalCols = Object.keys(analysisResult.statistics.categorical)
    const totalCols = analysisResult.summary.totalColumns
    const totalRows = analysisResult.summary.totalRows

    let problemType = 'unknown'
    if (targetVariable) {
      if (analysisResult.summary.columnTypes[targetVariable] === 'numerical') {
        problemType = 'regression'
      } else {
        problemType = 'classification'
      }
    }

    const prompt = `As an ML expert, recommend the top 3 models for this dataset:

Dataset Characteristics:
- Rows: ${totalRows.toLocaleString()}
- Columns: ${totalCols} (${numericalCols.length} numerical, ${categoricalCols.length} categorical)
- Problem Type: ${problemType || 'unsupervised'}
- Target Variable: ${targetVariable || 'None (unsupervised)'}

Data Quality:
- Missing Values: ${((Object.values(analysisResult.summary.missingValues).reduce((a, b) => a + b, 0) / (totalRows * totalCols)) * 100).toFixed(1)}%
- Strong Correlations: ${analysisResult.correlations.filter(c => Math.abs(c.correlation) > 0.7).length}

Recommend 3 specific ML models with:
1. Model name
2. Problem type it solves
3. Rationale for this dataset
4. Required preprocessing steps
5. Expected performance considerations
6. Important considerations

Format as JSON array if possible, otherwise plain text.`

    const result = await model.generateContent(prompt)
    const response = await result.response

    // Try to parse structured response
    try {
      const parsed = JSON.parse(response.text())
      return parsed.recommendations || []
    } catch {
      // Parse textual response
      return parseTextualRecommendations(response.text())
    }
  } catch (error) {
    console.warn('ML recommendation error:', error)
    return generateFallbackRecommendations(analysisResult, targetVariable)
  }
}

/**
 * Generate pattern insights from correlations
 */
export function detectPatterns(analysisResult: AnalysisResult): PatternInsight[] {
  const patterns: PatternInsight[] = []

  // Detect strong correlations
  const strongCorrelations = analysisResult.correlations.filter(c => Math.abs(c.correlation) > 0.7)
  if (strongCorrelations.length > 5) {
    patterns.push({
      pattern: 'Highly Correlated Features',
      description: `Found ${strongCorrelations.length} strong correlations between variables`,
      confidence: 0.9,
      implications: [
        'Consider feature selection to reduce dimensionality',
        'Monitor for multicollinearity in linear models',
        'Some features may be redundant'
      ]
    })
  }

  // Detect data quality issues
  const totalMissing = Object.values(analysisResult.summary.missingValues).reduce((a, b) => a + b, 0)
  const missingPercent = (totalMissing / (analysisResult.summary.totalRows * analysisResult.summary.totalColumns)) * 100

  if (missingPercent > 20) {
    patterns.push({
      pattern: 'Significant Missing Data',
      description: `${missingPercent.toFixed(1)}% of data is missing`,
      confidence: 1.0,
      implications: [
        'Imputation or removal strategies needed',
        'May impact model performance',
        'Consider missing data mechanisms'
      ]
    })
  }

  // Detect numerical vs categorical balance
  const numericalCount = Object.keys(analysisResult.statistics.numerical).length
  const categoricalCount = Object.keys(analysisResult.statistics.categorical).length

  if (numericalCount > categoricalCount * 3) {
    patterns.push({
      pattern: 'Numerical Feature Dominance',
      description: `Dataset has ${numericalCount} numerical vs ${categoricalCount} categorical features`,
      confidence: 0.8,
      implications: [
        'Well-suited for regression and neural networks',
        'Consider dimensionality reduction (PCA)',
        'Feature scaling will be important'
      ]
    })
  }

  return patterns
}

// Helper functions for fallback when Gemini API is unavailable

function generateFallbackNarrative(analysisResult: AnalysisResult): string {
  const insights = []

  insights.push(`This dataset contains ${analysisResult.summary.totalRows.toLocaleString()} rows and ${analysisResult.summary.totalColumns} columns, with a mix of numerical and categorical features.`)

  const totalMissing = Object.values(analysisResult.summary.missingValues).reduce((a, b) => a + b, 0)
  const missingPercent = (totalMissing / (analysisResult.summary.totalRows * analysisResult.summary.totalColumns)) * 100

  if (missingPercent > 20) {
    insights.push(`Data quality concerns exist with ${missingPercent.toFixed(1)}% missing values that should be addressed.`)
  } else if (missingPercent > 0) {
    insights.push(`Data quality is good with only ${missingPercent.toFixed(1)}% missing values.`)
  }

  const strongCorrelations = analysisResult.correlations.filter(c => Math.abs(c.correlation) > 0.7)
  if (strongCorrelations.length > 0) {
    insights.push(`Found ${strongCorrelations.length} strong correlations between variables that warrant further investigation.`)
  }

  return insights.join(' ')
}

function generateFallbackTransformations(columnName: string, stats: any): TransformationSuggestion[] {
  const suggestions: TransformationSuggestion[] = []

  // Check for potential skewness (if range is much larger than std)
  if (stats.std && stats.mean && (stats.max - stats.min) > 10 * stats.std) {
    suggestions.push({
      column: columnName,
      transformation: 'Log Transformation',
      reason: 'Reduces right-skew and compresses large values',
      code: `Math.log(${columnName} + 1)`
    })
  }

  // Check for different scales
  if (stats.std && stats.mean && stats.std / stats.mean > 0.5) {
    suggestions.push({
      column: columnName,
      transformation: 'Standardization (Z-score)',
      reason: 'Scales data to have mean=0, std=1 for better model convergence',
      code: `(${columnName} - ${stats.mean.toFixed(2)}) / ${stats.std.toFixed(2)}`
    })
  }

  return suggestions
}

function parseTextualSuggestions(text: string, columnName: string): TransformationSuggestion[] {
  const suggestions: TransformationSuggestion[] = []
  const lines = text.split('\n').filter(line => line.trim())

  lines.forEach(line => {
    if (line.toLowerCase().includes('log') || line.toLowerCase().includes('transform')) {
      suggestions.push({
        column: columnName,
        transformation: 'Log Transformation',
        reason: 'Based on AI analysis',
        code: `Math.log(${columnName} + 1)`
      })
    }
  })

  return suggestions
}

function parseTextualRecommendations(text: string): MLRecommendation[] {
  const recommendations: MLRecommendation[] = []

  // Simple text parsing for common model names
  if (text.toLowerCase().includes('random forest') || text.toLowerCase().includes('forest')) {
    recommendations.push({
      model: 'Random Forest',
      problemType: 'classification',
      rationale: 'Robust to outliers and non-linear relationships',
      preprocessing: ['Handle missing values', 'Encode categorical variables'],
      expectedPerformance: 'Good baseline model, interpretable feature importance',
      considerations: ['May overfit with many features', 'Requires tuning of hyperparameters']
    })
  }

  if (text.toLowerCase().includes('linear regression') || text.toLowerCase().includes('regression')) {
    recommendations.push({
      model: 'Linear Regression',
      problemType: 'regression',
      rationale: 'Simple, interpretable baseline for regression problems',
      preprocessing: ['Handle missing values', 'Scale numerical features', 'Encode categoricals'],
      expectedPerformance: 'Good for linear relationships, may underfit complex patterns',
      considerations: ['Assumes linearity', 'Sensitive to outliers', 'Requires feature scaling']
    })
  }

  if (text.toLowerCase().includes('gradient boosting') || text.toLowerCase().includes('xgboost')) {
    recommendations.push({
      model: 'Gradient Boosting (XGBoost)',
      problemType: 'classification',
      rationale: 'State-of-the-art performance for tabular data',
      preprocessing: ['Handle missing values', 'Encode categorical variables'],
      expectedPerformance: 'Excellent performance with proper tuning',
      considerations: ['Requires careful tuning', 'Can overfit on small datasets', 'Longer training time']
    })
  }

  return recommendations
}

function generateFallbackRecommendations(analysisResult: AnalysisResult, targetVariable?: string): MLRecommendation[] {
  const recommendations: MLRecommendation[] = []

  if (targetVariable) {
    const targetType = analysisResult.summary.columnTypes[targetVariable]

    if (targetType === 'numerical') {
      // Regression recommendations
      recommendations.push({
        model: 'Random Forest Regressor',
        problemType: 'regression',
        rationale: 'Handles non-linear relationships and mixed data types well',
        preprocessing: ['Handle missing values', 'Encode categorical variables'],
        expectedPerformance: 'Strong baseline with good interpretability',
        considerations: ['May require hyperparameter tuning', 'Feature importance available']
      })

      recommendations.push({
        model: 'Gradient Boosting Regressor',
        problemType: 'regression',
        rationale: 'Excellent performance on tabular regression tasks',
        preprocessing: ['Handle missing values', 'Feature scaling optional'],
        expectedPerformance: 'State-of-the-art with proper tuning',
        considerations: ['Longer training time', 'Requires careful hyperparameter tuning']
      })
    } else {
      // Classification recommendations
      recommendations.push({
        model: 'Random Forest Classifier',
        problemType: 'classification',
        rationale: 'Robust performer with built-in feature importance',
        preprocessing: ['Encode categorical variables', 'Handle missing values'],
        expectedPerformance: 'Strong baseline, interpretable',
        considerations: ['May overfit with many trees', 'Memory intensive']
      })

      recommendations.push({
        model: 'Gradient Boosting (XGBoost)',
        problemType: 'classification',
        rationale: 'Top performer for classification on tabular data',
        preprocessing: ['Encode categoricals', 'Handle missing values'],
        expectedPerformance: 'Excellent with proper tuning',
        considerations: ['Requires hyperparameter optimization', 'Training time longer']
      })
    }
  } else {
    // Unsupervised recommendations
    recommendations.push({
      model: 'K-Means Clustering',
      problemType: 'clustering',
      rationale: 'Simple and effective for finding groups in data',
      preprocessing: ['Feature scaling', 'Handle categorical variables'],
      expectedPerformance: 'Good baseline, requires determining optimal k',
      considerations: ['Assumes spherical clusters', 'Sensitive to outliers', 'Requires feature scaling']
    })
  }

  return recommendations
}

File: workers/ai_insights/gemini_client.py
"""
Gemini AI Client Module
Handles integration with Google Gemini API for generating insights.
"""
import os
import logging
from typing import Dict, Any, List
import google.generativeai as genai
from google.api_core.exceptions import GoogleAPIError

logger = logging.getLogger(__name__)


class GeminiClient:
    """Client for Google Gemini API."""

    def __init__(self):
        """Initialize Gemini client."""
        api_key = os.getenv('GEMINI_API_KEY')
        model_name = os.getenv('GEMINI_MODEL', 'gemini-1.5-flash')

        if not api_key:
            raise ValueError("GEMINI_API_KEY must be set in environment variables")

        genai.configure(api_key=api_key)
        self.model = genai.GenerativeModel(model_name)
        logger.info(f"Gemini client initialized with model: {model_name}")

    def generate_insights(self, analysis_data: Dict[str, Any], dataset_name: str) -> str:
        """
        Generate comprehensive AI insights from analysis results.

        Args:
            analysis_data: Complete analysis results
            dataset_name: Name of the dataset

        Returns:
            Generated insights text

        Raises:
            Exception: If API call fails
        """
        try:
            prompt = self._build_insight_prompt(analysis_data, dataset_name)

            logger.info("Sending request to Gemini API")
            response = self.model.generate_content(prompt)
            insights = response.text

            logger.info("Successfully generated insights")
            return insights

        except GoogleAPIError as e:
            logger.error(f"Gemini API error: {e}")
            raise Exception(f"Gemini API error: {str(e)}")
        except Exception as e:
            logger.error(f"Error generating insights: {e}")
            raise

    def _build_insight_prompt(self, analysis_data: Dict[str, Any], dataset_name: str) -> str:
        """
        Build comprehensive prompt for insight generation.

        Args:
            analysis_data: Analysis results
            dataset_name: Dataset name

        Returns:
            Prompt string
        """
        summary = analysis_data.get('summary', {})
        statistics = analysis_data.get('statistics', {})

        # Handle nested correlations structure
        correlations_data = analysis_data.get('correlations', [])
        if isinstance(correlations_data, dict) and 'correlations' in correlations_data:
            correlations = correlations_data['correlations']
        else:
            correlations = correlations_data if isinstance(correlations_data, list) else []

        quality = analysis_data.get('data_quality', {})
        ml_readiness = analysis_data.get('ml_readiness', {})

        # Build dataset overview
        overview = f"""Dataset: {dataset_name}
- Rows: {summary.get('total_rows', 'N/A'):,}
- Columns: {summary.get('total_columns', 'N/A')}
- Numerical: {summary.get('numerical_columns', 0)}
- Categorical: {summary.get('categorical_columns', 0)}"""

        # Build statistics summary
        stats_summary = self._build_statistics_summary(statistics)

        # Build correlations summary
        corr_summary = self._build_correlations_summary(correlations)

        # Build quality summary
        quality_summary = self._build_quality_summary(quality)

        # Build ML readiness summary
        ml_summary = self._build_ml_readiness_summary(ml_readiness)

        # Construct full prompt
        prompt = f"""You are an expert data scientist analyzing a dataset called "{dataset_name}".

{overview}

KEY STATISTICS:
{stats_summary}

CORRELATIONS:
{corr_summary}

DATA QUALITY:
{quality_summary}

ML READINESS:
{ml_summary}

Based on this analysis, provide a comprehensive 2-paragraph executive summary that includes:

1. First paragraph: Dataset overview and key patterns
   - Main characteristics and structure
   - Notable patterns in the data
   - Distribution of features

2. Second paragraph: Data quality assessment and recommendations
   - Critical data quality issues
   - Potential impact on analysis
   - Specific recommendations for improvement

Keep it professional, actionable, and under 200 words total. Focus on insights that would be valuable for a data scientist or analyst working with this dataset."""

        return prompt

    def _build_statistics_summary(self, statistics: Dict[str, Any]) -> str:
        """Build statistics summary for prompt."""
        lines = []

        numerical = statistics.get('numerical', {})
        for col, stats in list(numerical.items())[:5]:  # Limit to 5 columns
            lines.append(
                f"- {col}: mean={stats.get('mean', 0):.2f}, "
                f"std={stats.get('std', 0):.2f}, "
                f"range=[{stats.get('min', 0):.2f}, {stats.get('max', 0):.2f}]"
            )

        return "\n".join(lines) if lines else "No numerical statistics available"

    def _build_correlations_summary(self, correlations: List[Dict]) -> str:
        """Build correlations summary for prompt."""
        if not correlations:
            return "No significant correlations found"

        strong_correlations = [c for c in correlations if abs(c.get('correlation', 0)) > 0.7]
        lines = []

        for corr in strong_correlations[:5]:  # Limit to 5
            lines.append(
                f"- {corr.get('col1')} vs {corr.get('col2')}: "
                f"{corr.get('correlation', 0):.3f}"
            )

        return "\n".join(lines) if lines else "No strong correlations"

    def _build_quality_summary(self, quality: Dict[str, Any]) -> str:
        """Build data quality summary for prompt."""
        completeness = quality.get('completeness', {})
        uniqueness = quality.get('uniqueness', {})

        lines = [
            f"- Missing values: {completeness.get('missing_percentage', 0):.1f}%",
            f"- Duplicate rows: {uniqueness.get('duplicate_percentage', 0):.1f}%"
        ]

        return "\n".join(lines)

    def _build_ml_readiness_summary(self, ml_readiness: Dict[str, Any]) -> str:
        """Build ML readiness summary for prompt."""
        score = ml_readiness.get('overall_score', 0)
        level = ml_readiness.get('readiness_level', 'unknown')

        return f"ML Readiness Score: {score}/100 ({level})"

    def generate_column_insights(self, column_name: str, stats: Dict[str, Any]) -> str:
        """
        Generate insights for a specific column.

        Args:
            column_name: Column name
            stats: Column statistics

        Returns:
            Generated insights
        """
        try:
            prompt = f"""As a data scientist, provide a brief analysis of column "{column_name}" with these statistics:

Mean: {stats.get('mean', 'N/A')}
Median: {stats.get('median', 'N/A')}
Std Dev: {stats.get('std', 'N/A')}
Min: {stats.get('min', 'N/A')}
Max: {stats.get('max', 'N/A')}
Skewness: {stats.get('skewness', 'N/A')}

Provide 2-3 sentences about the distribution characteristics and any concerns."""

            response = self.model.generate_content(prompt)
            return response.text

        except Exception as e:
            logger.error(f"Error generating column insights: {e}")
            return "Unable to generate insights for this column"

    def generate_transformation_suggestions(self, column_name: str, stats: Dict[str, Any]) -> List[Dict[str, str]]:
        """
        Suggest data transformations for a column.

        Args:
            column_name: Column name
            stats: Column statistics

        Returns:
            List of transformation suggestions
        """
        try:
            prompt = f"""As a data scientist, recommend data transformations for column "{column_name}" with these statistics:

Mean: {stats.get('mean', 'N/A')}
Median: {stats.get('median', 'N/A')}
Std Dev: {stats.get('std', 'N/A')}
Min: {stats.get('min', 'N/A')}
Max: {stats.get('max', 'N/A')}
Skewness: {stats.get('skewness', 'N/A')}

Recommend 2-3 specific transformations to improve normality and ML model performance.
Format as JSON array:
[
  {{
    "transformation": "name",
    "reason": "why",
    "code": "example code"
  }}
]"""

            response = self.model.generate_content(prompt)

            # Try to parse JSON response
            import json
            try:
                suggestions = json.loads(response.text)
                return suggestions
            except:
                # Fallback to text parsing
                return self._parse_transformation_text(response.text, column_name)

        except Exception as e:
            logger.error(f"Error generating transformation suggestions: {e}")
            return []

    def _parse_transformation_text(self, text: str, column_name: str) -> List[Dict[str, str]]:
        """Parse transformation suggestions from text."""
        suggestions = []

        # Check for skewness and suggest log transform
        if 'log' in text.lower() or 'skew' in text.lower():
            suggestions.append({
                'transformation': 'Log Transformation',
                'reason': 'Reduces right-skew',
                'code': f'np.log({column_name} + 1)'
            })

        # Check for standardization
        if 'standard' in text.lower() or 'scale' in text.lower():
            suggestions.append({
                'transformation': 'Standardization',
                'reason': 'Scales to mean=0, std=1',
                'code': f'({column_name} - {column_name}.mean()) / {column_name}.std()'
            })

        return suggestions